### PR TITLE
*: only add union scan when dirty table is not empty

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

UnionScan has performance regression if it's empty, compared with scan data without it(Since it''s empty, we do can ignore it)

### What is changed and how it works?

If the dirty table is empty, we just don't add UnionScan.

### Check List 

Tests 

 - Existing Unit test

Code changes

 - Has exported function/method change

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2643,14 +2643,7 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
...
 
-	// If this SQL is executed in a non-readonly transaction, we need a
-	// "UnionScan" operator to read the modifications of former SQLs, which is
-	// buffered in tidb-server memory.
-	txn, err := b.ctx.Txn(false)
-	if err != nil {
-		return nil, err
-	}
-	if txn.Valid() && !txn.IsReadOnly() {
+	if b.ctx.HasDirtyContent(tableInfo.ID) {
```
Comment (lamxTyler) [inline planner/core/logical_plan_builder.go]: Should we consider partition table?

Comment (winoros) [reply planner/core/logical_plan_builder.go]: Good catch. Let me look into it.:joy:

**Comment by coocood** [comment]
LGTM

**Comment by winoros** [comment]
```
MySQL [tpch1]> begin;
Query OK, 0 rows affected (0.00 sec)

MySQL [tpch1]> insert into region values(111, 'fd', '');
Query OK, 1 row affected (0.01 sec)

MySQL [tpch1]> select count(*) from lineitem where l_returnflag = 'N';
+----------+
| count(*) |
+----------+
|  1287758 |
+----------+
1 row in set (1.21 sec)

MySQL [tpch1]> rollback;
Query OK, 0 rows affected (0.01 sec)

MySQL [tpch1]> select count(*) from lineitem where l_returnflag = 'N';
+----------+
| count(*) |
+----------+
|  1287758 |
+----------+
1 row in set (0.60 sec)
```

This is one case to show the improvement. When the data returned from the `DataSource` is large, we'll still get performance regression though there's no data in dirty table.

**Pull request merged**